lib/tasker.js

```diff
@@ -0,0 +1,150 @@
+/* global Zinnia */
+
+import { ActivityState } from './activity-state.js'
+import { encodeHex } from '../vendor/deno-deps.js'
+import { assertOkResponse, assertRedirectResponse } from './http-assertions.js'
+import { getRandomnessForSparkRound } from './drand-client.js'
+import { assertEquals, assertInstanceOf } from 'zinnia:assert'
+
+/** @typedef {{cid: string; minerId: string;}} RetrievalTask */
+/** @typedef {RetrievalTask & { key: string}} KeyedRetrievalTask */
+
+export class Tasker {
+  #lastRoundUrl
+  /** @type {Task[]} */
+  #remainingRoundTasks
+  #fetch
+  #activity
+
+  /**
+   * @param {object} args
+   * @param {globalThis.fetch} args.fetch
+   * @param {ActivityState} args.activityState
+   */
+  constructor ({
+    fetch = globalThis.fetch,
+    activityState = new ActivityState()
+  } = {}) {
+    this.#fetch = fetch
+    this.#activity = activityState
+
+    this.maxTasksPerRound = 360
+
+    // TODO: persist these two values across module restarts
+    this.#lastRoundUrl = 'unknown'
+    this.#remainingRoundTasks = []
+  }
+
+  /**
+   * @returns {Task | undefined}
+   */
+  async next () {
+    await this.#updateCurrentRound()
+    return this.#remainingRoundTasks.pop()
+  }
+
+  async #updateCurrentRound () {
+    console.log('Checking the current SPARK round...')
+    let res = await this.#fetch('https://api.filspark.com/rounds/current', {
+      method: 'GET',
+      headers: { 'Content-Type': 'application/json' },
+      redirect: 'manual',
+      signal: AbortSignal.timeout(10_000)
+    })
+
+    const roundUrl = await assertRedirectResponse(res, 'Failed to find the URL of the current SPARK round')
+    this.#activity.onHealthy()
+    if (roundUrl === this.#lastRoundUrl) {
+      console.log('Round did not change since the last iteration')
+      return
+    }
+
+    console.log('Fetching round details at location %s', roundUrl)
+    this.#lastRoundUrl = roundUrl
```

Comment:
should we move the state update to the end of the function, so that if there's an error we will try again next time?

Reply:
That's a great idea!